CXP-2188 [agent] add service discovery tests

gengnamstyle opened the pull request:
<!--
* Contributors are encouraged to read our [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Please fill the below sections if possible with relevant information or links.
-->
### What does this PR do?
- adds 2 e2e tests for service discovery (process collection enabled/disabled)
- enriches missing process data when process collection is disabled and service discovery is enabled
- adds language collector test to ensure language collector is disabled when `process_config.process_collection.use_wlm = true`
- updates clock usage to always use UTC time
- ensures language is still collected for service discovery if  `language_detection.enabled = false` (for process collection)

### Motivation
- service discovery data consolidation work is missing e2e tests

### Describe how you validated your changes
<!--
Validate your changes before merge, ensuring that:
* Your PR is tested by static / unit / integrations / e2e tests
* Your PR description details which e2e tests cover your changes, if any
* The PR description contains details of how you validated your changes. If you validated changes manually and not through automated tests, add context on why automated tests did not fit your changes validation.

If you want additional validation by a second person, you can ask reviewers to do it. Describe how to set up an environment for manual tests in the PR description. Manual validation is expected to happen on every commit before merge.

Any manual validation step should then map to an automated test. Manual validation should not substitute automation, minus exceptions not supported by test tooling yet.
-->
- run e2e tests manually 
- ci

### Possible Drawbacks / Trade-offs

### Additional Notes
<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

gengnamstyle committed:
update tests

gengnamstyle committed:
fix linting

gengnamstyle committed:
update e2e test config + change some assertions to require

vitkyrka commented on test/new-e2e/tests/discovery/testdata/config/system_probe_process_config.yaml:
```diff
@@ -0,0 +1,6 @@
+log_level: 'debug'
+system_probe_config:
+  process_config:
+    enabled: true
```
Yes, `system_probe_config.enabled` is set automatically by `pkg/system-probe/config/config.go` if any module is enabled.

gengnamstyle committed:
remove unnecessary e2e config

gengnamstyle committed:
fix when language detection is disabled

# Conflicts:
#	comp/core/workloadmeta/collectors/internal/process/process_service_collector_test.go

gengnamstyle committed:
update e2e test

gengnamstyle commented on comp/core/workloadmeta/collectors/internal/process/process_service_collector_test.go:
```diff
@@ -39,188 +42,6 @@ const (
-	assert.EventuallyWithT(t, func(collectT *assert.CollectT) {
-		for _, pid := range pids {
-			entity, err := store.GetProcess(pid)
-			assert.NoError(collectT, err, "PID %d should exist in store", pid)
-			assert.NotNil(collectT, entity, "PID %d should exist in store", pid)
-			// Process should exist but have no service data
-			assert.Nil(collectT, entity.Service, "PID %d should not have service data", pid)
-		}
-	}, 1*time.Second, 100*time.Millisecond)
-}
-
-func assertNoEntitiesForPids(t *testing.T, store workloadmetamock.Mock, pids []int32) {
-	if len(pids) == 0 {
-		return
-	}
-
-	assert.EventuallyWithT(t, func(collectT *assert.CollectT) {
-		for _, pid := range pids {
-			entity, err := store.GetProcess(pid)
-			assert.Error(collectT, err, "PID %d should not exist in store", pid)
-			assert.Nil(collectT, entity, "PID %d should exist in store", pid)
-		}
-	}, 1*time.Second, 100*time.Millisecond)
-}
-
-func assertProcessesExist(t *testing.T, store workloadmetamock.Mock, pids []int32) {
-	if len(pids) == 0 {
-		return
-	}
-
-	// Verify that processes exist (regardless of service data)
-	assert.EventuallyWithT(t, func(collectT *assert.CollectT) {
-		for _, pid := range pids {
-			entity, err := store.GetProcess(pid)
-			assert.NoError(collectT, err, "PID %d should exist in store", pid)
-			assert.NotNil(collectT, entity, "PID %d should exist in store", pid)
-		}
-	}, 1*time.Second, 100*time.Millisecond)
-}
-
```
moved to bottom to clean up testing file

robertjli replied:
Thought: How do we prevent people (or AI) from adding tests to the bottom of the file, _after_ these helpers?

gengnamstyle replied:
Yeah that's a good point, and I'm not too sure other than attempting to enforce this via reviews and perhaps more documentation on style guides?

robertjli replied:
Is having helpers at the bottom a recommended style somewhere? I don't have a huge preference on this but I do think it will be difficult to enforce. Not a blocking comment though.

robertjli commented on test/new-e2e/tests/discovery/linux_test.go:
```diff
@@ -161,6 +320,100 @@ func assertRunningCheck(t *assert.CollectT, remoteHost *components.RemoteHost, c
+	assert.NotContains(t, status.RunnerStats.Checks, check)
+}
+
+func assertProcessServiceDiscoveryData(t *testing.T, c *assert.CollectT, procs []*agentmodel.Process, expectedLanguage agentmodel.Language, expectedPortInfo *agentmodel.PortInfo, expectedServiceDiscovery *agentmodel.ServiceDiscovery) {
```
Thought: This function is quite long and was hard to read for me. You can consider breaking it up into helpers for each check that you do.

robertjli commented on test/new-e2e/tests/discovery/linux_test.go:
```diff
@@ -161,6 +320,100 @@ func assertRunningCheck(t *assert.CollectT, remoteHost *components.RemoteHost, c
+				continue
+			}
+			// tracer metadata contains a uuid (TracerMetadata.RuntimeID), so we do a manual comparison here
+			//correctServiceDiscovery = correctServiceDiscovery && assert.ElementsMatch(noopt{}, expectedServiceDiscovery.TracerMetadata, proc.ServiceDiscovery.TracerMetadata)
```
Is this supposed to be uncommented or deleted?

robertjli commented on test/new-e2e/tests/discovery/linux_test.go:
```diff
@@ -161,6 +320,100 @@ func assertRunningCheck(t *assert.CollectT, remoteHost *components.RemoteHost, c
+}
+
+// noopt is an empty struct that implements the Errorf method of the error interface so we can reuse testify functions
+type noopt struct{}
```
I realize you used `go-cmp` for the `TracerMetadata` comparisons and sorted those slices—in that case, why not do that for all of them and forgo the hacky use of `assert.ElementsMatch`?

gengnamstyle committed:
cleanup

gengnamstyle commented on comp/core/workloadmeta/collectors/internal/process/process_collector_test.go:
```diff
@@ -119,6 +119,8 @@ func TestBasicCreatedProcessesCollection(t *testing.T) {
 				}
 			}, time.Second, time.Millisecond*100)
+			// needed to reset the global telemetry registry as the start function registers a new gauge when discovery is enabled for each run
+			telemetry.GetCompatComponent().Reset()
```
Yup i agree! I liked the cleanup, so I implemented that. 🙏

gengnamstyle committed:
fix test condition panic

robertjli commented on comp/core/workloadmeta/collectors/internal/process/process_collector_test.go:
```diff
@@ -16,17 +16,18 @@ import (
+	"github.com/DataDog/datadog-agent/pkg/errors"
+	"github.com/DataDog/datadog-agent/pkg/languagedetection/languagemodels"
+	"github.com/DataDog/datadog-agent/pkg/telemetry"
+
```
Chore: combine these two import blocks as they are both all datadog-agent imports

robertjli commented on comp/core/workloadmeta/collectors/internal/process/process_service_collector_test.go:
```diff
@@ -15,21 +15,22 @@ import (
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/servicediscovery/language"
+	"github.com/DataDog/datadog-agent/pkg/languagedetection/languagemodels"
+	"github.com/DataDog/datadog-agent/pkg/telemetry"
+
```
Chore: combine these two import blocks as they are both all datadog-agent imports

robertjli commented on test/new-e2e/tests/discovery/linux_test.go:
```diff
@@ -329,90 +329,87 @@ func assertNotRunningCheck(t *assert.CollectT, remoteHost *components.RemoteHost
-// noopt is an empty struct that implements the Errorf method of the error interface so we can reuse testify functions
-type noopt struct{}
+func matchingServiceNames(expectedServiceNames []*agentmodel.ServiceName, actualServiceNames []*agentmodel.ServiceName) bool {
+	// tracer metadata contains a uuid (TracerMetadata.RuntimeID), so we ignore it
```
Comment is not relevant, seems to be mistakenly pasted?

gengnamstyle replied:
removed thanks

robertjli commented on test/new-e2e/tests/discovery/linux_test.go:
```diff
@@ -329,90 +329,87 @@ func assertNotRunningCheck(t *assert.CollectT, remoteHost *components.RemoteHost
+		if !matchingServiceNames(expectedServiceDiscovery.AdditionalGeneratedNames, proc.ServiceDiscovery.AdditionalGeneratedNames) {
+			continue
 		}
+		assert.True(t, true)
```
Remove, this always passes

robertjli commented on test/new-e2e/tests/discovery/linux_test.go:
```diff
@@ -155,12 +314,103 @@ func assertCollectorStatusFromJSON(t *assert.CollectT, statusOutput, check strin
+			continue
+		}
+
+		if !matchingServiceNames([]*agentmodel.ServiceName{expectedServiceDiscovery.GeneratedServiceName}, []*agentmodel.ServiceName{proc.ServiceDiscovery.GeneratedServiceName}) {
```
Optional: you could create another helper to remove the slice creations and make this invocation a bit cleaner

```
func matchingServiceName(a, b *agentmodel.ServiceName) bool {
	return matchingServiceNames([]*agentmodel.ServiceName{a}, []*agentmodel.ServiceName{b})
}
```

gengnamstyle replied:
created thanks for the suggestion

gengnamstyle committed:
nits + fix assertion test

gengnamstyle committed:
e2e test helper fixed

gengnamstyle committed:
clean up test reset code

robertjli committed:
Merge branch 'main' into matthew.geng/CXP-2188-new-process-collection-e2e-tests

robertjli commented:
/merge
